src/spaces/discrete.jl

```diff
@@ -78,13 +78,29 @@ function has_empty_positions(model::ABM{<:DiscreteSpace})
 end
 
 """
-    random_empty(model::ABM{<:DiscreteSpace})
+    random_empty(model::ABM{<:DiscreteSpace}, cutoff = 0.998)
 Return a random position without any agents, or `nothing` if no such positions exist.
-"""
-function random_empty(model::ABM{<:DiscreteSpace})
-    empty = empty_positions(model)
-    isempty(empty) && return nothing
-    rand(model.rng, collect(empty))
+`cutoff` switches the search algorithm from probabalistic to a filter.
+"""
+function random_empty(model::ABM{<:DiscreteSpace}, cutoff = 0.998)
+    # This switch assumes the worst case (for this algorithm) of one
+    # agent per position, which is not true in general but is appropriate
+    # here.
+    if nagents(model) / prod(size(model.space.s)) < cutoff
+        # 0.998 has been benchmarked as a performant branching point
```

Comment:
Is the 0.998 hardware dependent maybe?

Reply:
I don't think it'll be hardware dependent, but could be space type / dimension dependent perhaps. The 0.998 seemed to scale pretty well for 2D grid space at least, so it'll be good to take as a default.

Before doing any merging I'll be sense checking graphs and higher dimensional grid spaces.